src/java/org/apache/cassandra/index/sai/disk/v1/TermsReader.java

```diff
@@ -262,43 +260,46 @@ public PostingList execute()
+
+            do
             {
-                do
+                Pair<ByteComparable, Long> nextTriePair = reader.next();
+                ByteSource mapEntry = nextTriePair.left.asComparableBytes(ByteComparable.Version.OSS41);
+                long postingsOffset = nextTriePair.right;
+                byte[] nextBytes = ByteSourceInverse.readBytes(mapEntry);
+
+                if (exp.isSatisfiedBy(ByteBuffer.wrap(nextBytes)))
                 {
-                    Pair<ByteComparable, Long> nextTriePair = triePairs.next();
-                    ByteSource mapEntry = nextTriePair.left.asComparableBytes(ByteComparable.Version.OSS41);
-                    long postingsOffset = nextTriePair.right;
-                    byte[] nextBytes = ByteSourceInverse.readBytes(mapEntry);
-                    if (exp.isSatisfiedBy(ByteBuffer.wrap(nextBytes)))
-                    {
-                        var blocksSummary = new PostingsReader.BlocksSummary(postingsSummaryInput, postingsOffset, PostingsReader.InputCloser.NOOP);
-                        @SuppressWarnings("resource")
-                        var currentReader = new PostingsReader(postingsInput,
-                                                               blocksSummary,
-                                                               listener.postingListEventListener(),
-                                                               PostingsReader.InputCloser.NOOP);
-                        if (heap == null)
-                            heap = new LongHeap((int) currentReader.size());
-                        while (true)
-                        {
-                            long nextPosting = currentReader.nextPosting();
-                            if (nextPosting == PostingList.END_OF_STREAM)
-                                break;
-                            heap.push(nextPosting);
-                        }
-                    }
-                } while (triePairs.hasNext());
-                return heap != null ? heap : new LongHeap(1);
-            }
+                    var blocksSummary = new PostingsReader.BlocksSummary(postingsSummaryInput, postingsOffset, PostingsReader.InputCloser.NOOP);
+                    var currentReader = new PostingsReader(postingsInput,
+                                                           blocksSummary,
+                                                           listener.postingListEventListener(),
+                                                           PostingsReader.InputCloser.NOOP);
+
+                    if (currentReader.size() > 0)
+                        postingLists.add(new PostingList.PeekablePostingList(currentReader));
+                    else
+                        FileUtils.close(currentReader);
+                }
+            } while (reader.hasNext());
+
+            return postingLists.isEmpty()
+                   ? PostingList.EMPTY.onClose(onClose)
```

Comment:
Nit: is it worth calling `onClose` eagerly here, as the comment above indicates?

Reply:
Ah, the part saying ` or immediately if it is empty` in the comment above is wrong, it's a remnant of a previous implementation. The current call to `onClose` will close the inputs when the empty posting list is closed.

We can either update the comment or keep it and eagerly close the inputs here with:
```java
if (postingLists.isEmpty())
{
    onClose.close();
    return PostingList.EMPTY;
}
return MergePostingList.merge(postingLists, onClose);
```
Or maybe just:
```java
if (postingLists.isEmpty())
{
    FileUtils.close(postingsInput, postingsSummaryInput);
    return PostingList.EMPTY;
}
return MergePostingList.merge(postingLists, () -> FileUtils.close(postingsInput, postingsSummaryInput));
```
I think I prefer to keep the current code and update the comment so things are closed at the same moment (when calling `PostingList#close()` with both empty and not empty iterators. WDYT?

Reply:
My assumption is that it is more efficient to close the files eagerly, so if we can do that without too much effort, I think we should. What do you think about updating the below implementation

```java
    default PostingList onClose(Closeable onClose)
    {
        return onClose == null ? this : new PostingListWithOnClose(this, onClose);
    }
```

so that when `this.isEmpty()` is true, we return call `onClose` and return `PostingList.EMPTY`.

Reply:
Agree. I have changed `PostingList#onClose` to immediately close the `Closeable` if the list is empty. I have updated the method's JavaDoc to make clear that the closing isn't necessarily done exactly during `#close()`, but it will be done when the `#close()` call is finished.

The `MergePostingList#merge` methods rejected an empty list but special-cased a single-element list. This is a bit asymmetrical, so I have changed them to special-casing both cases, so they also deal with empty lists. This makes things a bit easier for callers.

I have also noticed that `MergePostingList#merge` was missing to include the `Closeable` argument if there was a single posting list. This is probably a bug, maybe hit by `BKDReader.Intersection#mergePostings`. So I have removed the `Closeable` logic from `MergePostingList`,  and let the callers use the inherited `onClose` method. 

This IMO simplifies things a bit and prevents forgetting to close resources in special cases. wdyt?

Reply:
Looks good to me, thanks for changing it.